### PR TITLE
Add conformance tests for `member-access-expr`

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/32855

### CODE REVIEW

File: conformance/lang/expressions/member-access-expr/array_member_access_expr.balt [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+function init() {
+    io:println(getInt([1, 2, 3, 4, 5])); // @output 5
+}
+
+function getInt(int[] arr) returns int {
+    return arr[4];
+}
+
+Test-Case: output
+Description: Test assigning a value to member access expression.
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/array_member_access_expr.balt:517]: Also, these rules are described under [assigment-stmt](https://ballerina.io/spec/lang/master/#assignment-stmt) where `lvexpr` is a `member-access-lvexpr`.

File: conformance/lang/expressions/member-access-expr/array_member_access_expr.balt [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+function init() {
+    io:println(getInt()); // @output 5
+}
+
+function getInt(int a = arr[4]) returns int {
+    return a;
+}
+
+Test-Case: output
+Description: Test member access expression with container expression of basic type list as expression in
```
[the remark this answers is not in the document]
Comment (lasinicl) [reply conformance/lang/expressions/member-access-expr/array_member_access_expr.balt:555]: Didn't remove.

File: conformance/lang/expressions/member-access-expr/array_member_access_expr.balt [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+
+    IntType2 i4 = 5;
+    io:println(arr[i4]); // @output 60
+
+    io:println(arr[5]); // @output 60
+}
+
+Test-Case: output
+Description: Test whether the static type of key expression is subtype of int when the basic type of the
+             container expression is an array.
```
[the remark this answers is not in the document]
Comment (pcnfernando) [reply conformance/lang/expressions/member-access-expr/array_member_access_expr.balt:41]: Check other places as well

File: conformance/lang/expressions/member-access-expr/array_member_access_expr.balt [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+Fail-Issue: ballerina-platform/ballerina-lang#33346
```
Comment (pcnfernando) [inline conformance/lang/expressions/member-access-expr/array_member_access_expr.balt:42]: https://github.com/ballerina-platform/ballerina-lang/issues/33346 is fixed on 2201.0.x

File: conformance/lang/expressions/member-access-expr/array_member_access_expr.balt [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+                {a: 2},
+                {a: 3, b: "DEF"}
+            ],
+        table [
+                {a: 4},
+                {a: 5, b: "GHI"},
+                {a: 6}
+            ]
+    ];
+    io:println(tableArr[1]); // @output [{"a":4},{"a":5,"b":"GHI"},{"a":6}]
```
Comment (pcnfernando) [inline conformance/lang/expressions/member-access-expr/array_member_access_expr.balt:130]: Shall we test the result type as well

Comment (pcnfernando) [reply conformance/lang/expressions/member-access-expr/array_member_access_expr.balt:130]: Tested at 205

File: conformance/lang/expressions/member-access-expr/record_member_aceess_expr.balt [hunk 6]
```diff
@@ -0,0 +1,716 @@
+Test-Case: parser-error
```
[the remark this answers is not in the document]
Comment (lasinicl) [reply conformance/lang/expressions/member-access-expr/record_member_aceess_expr.balt:1]: Added

File: conformance/lang/expressions/member-access-expr/array_member_access_expr.balt [hunk 7]
```diff
@@ -0,0 +1,594 @@
+Test-Case: parser-error
```
[the remark this answers is not in the document]
Comment (lasinicl) [reply conformance/lang/expressions/member-access-expr/array_member_access_expr.balt]: Have added some tests like this in each file.
https://github.com/ballerina-platform/ballerina-spec/pull/1044/files#diff-c08eb78533280c6562ac87546406ccab92b6a0343f521533ad53336e09d48ae6R601-R609

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 8]
```diff
@@ -0,0 +1,639 @@
...
+
+Test-Case: output
+Description: Test query expression as container expression of basic type list.
+Labels: array-type, int, list-constructor-expr, member-access-expr, multiplicative-expr, query-expr
+
+function init() {
+    int[5] arr = [1, 2, 3, 4, 5];
+    io:println((from int i in arr select i * 10)[4]); // @output 50
+}
+
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:348]: Can also add a test with container expression being a list constructor. e.g. `[1, 2, 3][0]`

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 9]
```diff
@@ -0,0 +1,639 @@
...
+Labels: array-type, int, list-constructor-expr, member-access-expr
+
+function init() {
+    int[3] arr = [1, 2, 3];
+    arr[1] = 12;
+    io:println(arr[1]); // @output 12
+}
+
+Test-Case: error
+Description: Test using a value greater than the size of a sealed array as the key.
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:605]: Let's avoid the term sealed array as suggested earlier.

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 10]
```diff
@@ -0,0 +1,639 @@
...
+}
+
+Test-Case: error
+Description: Test invalid member access using multi key expression.
+Labels: array-type, int, list-constructor-expr, member-access-expr
+
+function errorFunction() {
+    int[3] arr = [1, 2, 3];
+    _ = arr[1, 2]; // @error member access with multi-key expression is not allowed with array
+}
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:639]: Also a test with member access at module-level.
```bal
int[3] arr = [1, 2, 3];
int x = arr[1];

function init() {
    io:println(x);
}
```

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 11]
```diff
@@ -0,0 +1,639 @@
...
+             the container expression is an array.
+Labels: array-type, byte, int, int:Signed16, int:Signed32, int:Signed8, list-constructor-expr, member-access-expr,
+        module-type-defn, union-type
+
+type IntType1 int:Signed8|int:Signed16;
+
+function init() {
+    int[8] arr = [10, 20, 30, 40, 50, 60, 70, 80];
+
+    int:Signed8 i1 = 1;
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:74]: A test with an Unsigned built-in subtype also.

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 12]
```diff
@@ -0,0 +1,639 @@
...
+    io:println(arr[i3]); // @output 40
+
+    IntType1 i4 = 5;
+    io:println(arr[i4]); // @output 60
+}
+
+Test-Case: error
+Description: Test key expression with invalid static type.
+Labels: any, array-type, DecimalFloatingPointNumber, float, int, list-constructor-expr, member-access-expr, string,
+        union-type
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:90]: labels: module-type-defn, module-const-decl, singleton-type

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 13]
```diff
@@ -0,0 +1,639 @@
...
+    _ = arr[e]; // @error expected a value of subtype int as key
+
+    _ = arr[2.0]; // @error expected a value of subtype int as key
+}
+
+Test-Case: output
+Description: Test the result of evaluating a member access expression when basic type of the evaluated value of the
+             container expression is an array.
+Labels: array-type, int, list-constructor-expr, mapping-constructor-expr, member-access-expr, module-type-defn,
+        readonly-field, record-type, string, table-constructor-expr, table-type, union-type
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:122]: also missing labels: table-type-key-specifier

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 14]
```diff
@@ -0,0 +1,639 @@
...
+function init() {
+    int[8] intArr = [10, 20, 30, 40, 50, 60, 70, 80];
+
+    _ = intArr[10]; // @panic index out of range
+}
+
+Test-Case: output
+Description: Test the static type of member access expression when the container expression is of basic type list.
+Labels: array-type, int, list-constructor-expr, mapping-constructor-expr, member-access-expr, module-type-defn,
+        record-type, string, table-constructor-expr, table-type, union-type
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:194]: labels: record-type-readonly-field, table-type-key-specifier

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 15]
```diff
@@ -0,0 +1,639 @@
...
+    int[] anotherArr = [arr[3], arr[0], arr[4]];
+
+    io:println(anotherArr[1]); // @output 1
+}
+
+Test-Case: output
+Description: Test member access expression with container expression of basic type list as default expression
+             for a field.
+Labels: array-type, int, list-constructor-expr, mapping-constructor-expr, member-access-expr, module-init-var-decl,
+        module-type-defn, readonly-type, record-type
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:552]: labels: field-access-expr, intersection-type

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 16]
```diff
@@ -0,0 +1,639 @@
...
+function init() {
+    Record rec = {};
+    io:println(rec.a); // @output 3
+}
+
+Test-Case: output
+Description: Test member access expression with container expression of basic type list as defaultable
+             function parameter.
+Labels: array-type, defaultable-param, function-call-expr, int, list-constructor-expr, member-access-expr,
+        module-init-var-decl, readonly-type
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:569]: intersection-type

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 17]
```diff
@@ -0,0 +1,639 @@
...
+Labels: array-type, int, list-constructor-expr, member-access-expr, optional-type
+
+function errorFunction() {
+    int[3]? arr = [1, 2, 3];
+    _ = arr[1]; // @error int[]? does not support member access
+}
+
+Test-Case: error
+Description: Test assigning an invalid value to member access expression.
+Labels: array-type, int, list-constructor-expr, member-access-expr, string
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:625]: member-access-lvexpr

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt (hunk 10 above)
Comment (lochana-chathura) [reply conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:639]: my bad. there is a test case.

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 18]
```diff
@@ -0,0 +1,1219 @@
...
+    io:println(comment[1].toBalString()); // @output xml``
+    // k > length(c)
+    io:println(comment[4].toBalString()); // @output xml``
+
+    xml:Text text = xml `Some Text`;
+    // k == length(c)
+    io:println(text[1].toBalString()); // @output xml``
+    // k > length(c)
+    io:println(text[4].toBalString()); // @output xml``
+}
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:661]: Spec says,

> if it is xml, and k is < 0, then the evaluation completes abruptly with a panic; 

Do we have a test covering that? Seems in the implementation we do not panic for k<0, instead we return an empty XML value.

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 19]
```diff
@@ -0,0 +1,1219 @@
...
+
+function errorFunction() {
+    xml e = xml `<name>Enid Blyton</name><name>Dan Brown</name>`;
+    string _ = e[1]; // @error @error expected string, found xml
+}
+
+Test-Case: output
+Description: Test assigning the result of evaluating a member access expression to var when basic type of the evaluated
+             value of the container expression is an xml.
+Labels: is-expr, member-access-expr, xml, xml:Comment, xml:Element, xml:length, xml:ProcessingInstruction, xml:Text
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:790]:
```suggestion
Labels: is-expr, member-access-expr, xml, xml:Comment, xml:Element, xml:length, xml:ProcessingInstruction, xml:Text, var
```

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 20]
```diff
@@ -0,0 +1,1219 @@
...
+Labels: additive-expr, member-access-expr, string, xml
+
+function init() {
+    io:println((xml `<name>Enid Blyton</name>` + xml `<name>Dan Brown</name>`)[1]); // @output <name>Dan Brown</name>
+    io:println((xml `Text1` + "Text2")[0]); // @output Text1Text2
+}
+
+Test-Case: output
+Description: Test conditional expression as container expression whose static type is xml.
+Labels: conditional-expr, is-expr, member-access-expr, nil-literal, optional-type, xml
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:889]:
```suggestion
Labels: ternary-conditional-expr, is-expr, member-access-expr, nil-literal, optional-type, xml
```

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 21]
```diff
@@ -0,0 +1,1219 @@
...
+
+function init() returns error? {
+    xml? msg = ();
+    io:println((msg is xml ? xml `<name>Enid Blyton</name>` : xml `<name>Dan Brown</name>`)[0]); // @output <name>Dan Brown</name>
+}
+
+Test-Case: output
+Description: Test checking expression as container expression whose static type is xml.
+Labels: check, checkpanic, error, member-access-expr, union-type, xml
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:898]:
```suggestion
Labels: check, checkpanic, error-type, member-access-expr, union-type, xml
```
Here and other places we have used `error`

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 22]
```diff
@@ -0,0 +1,1219 @@
...
+
+function init() returns error? {
+    xml|error val = xml `<name>Enid Blyton</name>`;
+    io:println((check val)[0]); // @output <name>Enid Blyton</name>
+    io:println((checkpanic val)[0]); // @output <name>Enid Blyton</name>
+}
+
+Test-Case: output
+Description: Test query expression as container expression whose static type is xml.
+Labels: array-type, list-constructor-expr, member-access-expr, query-expr, xml
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:908]:
```suggestion
Labels: array-type, list-constructor-expr, member-access-expr, query-expr, xml, var
```

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 23]
```diff
@@ -0,0 +1,1219 @@
...
+
+function init() returns error? {
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:910]: do we need `returns error?` for this test, two above and below.

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 24]
```diff
@@ -0,0 +1,1219 @@
...
+    io:println(val[getPosition()]); // @output Some Text
+}
+
+function getPosition() returns int {
+    return 2;
+}
+
+Test-Case: output
+Description: Test method call as key expression with container expression whose static type is xml.
+Labels: explicit-new-expr, int, member-access-expr, method-call-expr, module-class-defn, xml
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:981]: additive-expr, unary-minus

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 25]
```diff
@@ -0,0 +1,1219 @@
...
+    io:println(getMember(xml `<name>Enid Blyton</name><name>Dan Brown</name>`)); // @output <name>Dan Brown</name>
+}
+
+function getMember(xml member) returns xml {
+    return member[1];
+}
+
+Test-Case: output
+Description: Test accessing members of a member access expression.
+Labels: member-access-expr
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:1167]: shall we complete the missing labels?

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 26]
```diff
@@ -0,0 +1,1219 @@
...
+    [int, [xml...]] t1 = [1, [xml `<name>Enid Blyton</name><name>Dan Brown</name>`, xml `Some Text`, xml `<!--Comment-->`]];
+    io:println(t1[1][0][1][0][0]); // @output <name>Dan Brown</name>
+
+    Record rec = {a: [[xml `<name>Enid Blyton</name><name>Dan Brown</name>`, xml `Some Text`], [xml `<name>Enid Blyton</name><name>Dan Brown</name>`, xml `<!--Comment-->`], [xml `<name>Dan Brown</name>`]]};
+    io:println(rec["a"][1][0][1][0][0]); // @output <name>Dan Brown</name>
+}
+
+Test-Case: error
+Description: Test assigning a value to member access expression.
+Labels: member-access-expr, union-type, xml, xml:Comment, xml:Element, xml:Text
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:1189]:
```suggestion
Labels: member-access-lvexpr, union-type, xml, xml:Comment, xml:Element, xml:Text
```

File: conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt [hunk 27]
```diff
@@ -0,0 +1,1219 @@
...
+    xml:Element e2 = xml `<name>Enid Blyton</name>`;
+    e[0] = xml `<name>Nicholas Sparks</name>`; // @error cannot update an xml sequence
+
+    xml:Text|xml:Comment e3 = xml `Text`;
+    e[0] = xml `Another Text`; // @error cannot update an xml sequence
+}
+
+Test-Case: error
+Description: Test accessing members of an invalid structure.
+Fail-Issue: ballerina-platform/ballerina-lang#35032
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/xml_member_access_expr.balt:1204]: The issue is closed. Shall we point to the currently opened issue?

File: conformance/lang/expressions/member-access-expr/record_member_aceess_expr.balt [hunk 28]
```diff
@@ -0,0 +1,879 @@
...
+function init() {
+    Person p1 = {id: 5, name: "Mike"};
+
+    p1["id"] = 7;
+    io:println(p1["id"]); // @output 7
+}
+
+Test-Case: error
+Description: Test accessing members of an invalid structure.
+Labels: error-type, mapping-constructor-expr, member-access-expr, module-type-defn, record-type
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/record_member_aceess_expr.balt:838]:
```suggestion
Labels: error-type, mapping-constructor-expr, member-access-expr, module-type-defn, record-type, union-type
```

File: conformance/lang/expressions/member-access-expr/table_member_access_expr.balt [hunk 29]
```diff
@@ -0,0 +1,1153 @@
...
+            {id: 1, name: "Mike", age: 20},
+            {id: 2, name: "Tobi", age: 12},
+            {id: 3, name: "Frank", age: 15}
+        ];
+
+}
+
+Test-Case: output
+Description: Test method call expression as container expression of basic type table.
+Labels: explicit-new-expr, int, member-access-expr, method-call-expr, module-class-defn, module-type-defn,
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/table_member_access_expr.balt:450]:
```suggestion
Labels: implicit-new-expr, int, member-access-expr, method-call-expr, module-class-defn, module-type-defn,
```

File: conformance/lang/expressions/member-access-expr/table_member_access_expr.balt [hunk 30]
```diff
@@ -0,0 +1,1153 @@
...
+    return 3;
+}
+
+function getName() returns string {
+    return "Frank";
+}
+
+Test-Case: output
+Description: Test method call in key expression with container expression of basic type list.
+Labels: explicit-new-expr, int, member-access-expr, method-call-expr, module-class-defn, module-type-defn,
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/table_member_access_expr.balt:690]:
```suggestion
Labels: implicit-new-expr, int, member-access-expr, method-call-expr, module-class-defn, module-type-defn,
```

File: conformance/lang/expressions/member-access-expr/table_member_access_expr.balt [hunk 31]
```diff
@@ -0,0 +1,1153 @@
...
+type PersonTable table<Person> key(id, name);
+
+function errorFunction() {
+    PersonTable tbl = table [
+            {id: 1, name: "Mike", age: 15},
+            {id: 2, name: "Tobi", age: 16},
+            {id: 3, name: "Frank", age: 14}
+        ];
+
+    _ = tbl["1", "Mike"]; // @error invalid key constraint
```
Comment (lochana-chathura) [inline conformance/lang/expressions/member-access-expr/table_member_access_expr.balt:124]:
```suggestion
    _ = tbl["1", "Mike"]; // @error invalid key constraint, key constraint expected with type '[int,string]'
```

**Review by lochana-chathura — COMMENTED** [review]
Tests LGTM. Added a few label related suggestions.

**Review by lochana-chathura — APPROVED** [review]

File: conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt [hunk 32]
```diff
@@ -0,0 +1,663 @@
...
+    }
+}
+
+function init() {
+    Array arr = new Array();
+    io:println(arr.getArr()[2]); // @output 30
+}
+
+Test-Case: output
+Description: Test let expression expression as container expression of basic type list.
```
Comment (pcnfernando) [inline conformance/lang/expressions/member-access-expr/fixed_length_array_member_access_expr.balt:326]:
```suggestion
Description: Test let expression as container expression of basic type list.
```

File: conformance/lang/expressions/member-access-expr/non_fixed_length_array_member_access_expr.balt [hunk 33]
```diff
@@ -0,0 +1,641 @@
...
+    }
+}
+
+function init() {
+    Array arr = new Array();
+    io:println(arr.getArr()[2]); // @output 30
+}
+
+Test-Case: output
+Description: Test let expression expression as container expression of basic type list.
```
Comment (pcnfernando) [inline conformance/lang/expressions/member-access-expr/non_fixed_length_array_member_access_expr.balt:317]:
```suggestion
Description: Test let expression as container expression of basic type list.
```

File: conformance/lang/expressions/member-access-expr/map_member_access_expr.balt [hunk 34]
```diff
@@ -0,0 +1,505 @@
...
+    }
+}
+
+function init() {
+    Map m = new;
+    io:println(m.getMap()["name"]); // @output John
+}
+
+Test-Case: output
+Description: Test let expression expression as container expression of type map.
```
Comment (pcnfernando) [inline conformance/lang/expressions/member-access-expr/map_member_access_expr.balt:231]: Check and fix other places as well

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**